src/Mod/Draft/draftfunctions/downgrade.py

```diff
@@ -0,0 +1,274 @@
+# *   Copyright (c) 2009, 2010 Yorik van Havre <yorik@uncreated.net>        *
+# *   Copyright (c) 2009, 2010 Ken Cline <cline@frii.com>                   *
+# *   Copyright (c) 2020 FreeCAD Developers                                 *
+# *                                                                         *
+# *   This program is free software; you can redistribute it and/or modify  *
+# *   it under the terms of the GNU Lesser General Public License (LGPL)    *
+# *   as published by the Free Software Foundation; either version 2 of     *
+# *   the License, or (at your option) any later version.                   *
+# *   for detail see the LICENCE text file.                                 *
+# *                                                                         *
+# *   This program is distributed in the hope that it will be useful,       *
+# *   but WITHOUT ANY WARRANTY; without even the implied warranty of        *
+# *   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the         *
+# *   GNU Library General Public License for more details.                  *
+# *                                                                         *
+# *   You should have received a copy of the GNU Library General Public     *
+# *   License along with this program; if not, write to the Free Software   *
+# *   Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  *
+# *   USA                                                                   *
+# *                                                                         *
+# ***************************************************************************
+"""This module provides the code for Draft offset function.
+"""
+## @package offset
+# \ingroup DRAFT
+# \brief This module provides the code for Draft offset function.
+
+import FreeCAD as App
+
+import draftutils.gui_utils as gui_utils
+import draftutils.utils as utils
+
+from draftutils.translate import _tr
+
+from draftutils.utils import shapify
+from draftfunctions.cut import cut
+
+
+def downgrade(objects, delete=False, force=None):
+    """downgrade(objects,delete=False,force=None)
+    
+    Downgrade the given object(s) (can be an object or a list of objects).
+
+    Parameters
+    ----------
+    objects :
+
+    delete : bool
+        If delete is True, old objects are deleted.
+
+    force : string
+        The force attribute can be used to force a certain way of downgrading.
+        It can be: explode, shapify, subtr, splitFaces, cut2, getWire,
+        splitWires, splitCompounds.
+    
+    Return
+    ----------
+        Returns a dictionary containing two lists, a list of new objects and a
+        list of objects to be deleted
+    """
+
+    import Part
+    import DraftGeomUtils
+
+    if not isinstance(objects,list):
+        objects = [objects]
+
+    global deleteList, addList
+    deleteList = []
+    addList = []
+
+    # actions definitions
+
+    def explode(obj):
```

Comment:
By the way, it's not good to define functions inside functions. So, ideally, this should be refactored so that the function is outside.

Reply:
this function is huge so it will benefit for sure if we split something. But i observe it's better if we do it later, after it is moved to its own module.